infrastructure/comms/src/connection/message.rs

```diff
@@ -23,55 +23,78 @@
 use crate::peer_manager::node_id::*;
 use bitflags::*;
 use derive_error::Error;
-use serde_derive::{Deserialize, Serialize};
+use rmp_serde;
+use serde::{Deserialize, Serialize};
 use std::convert::TryFrom;
 use tari_crypto::keys::PublicKey;
 
+/// Represents a single message frame.
+pub type Frame = Vec<u8>;
+/// Represents a collection of frames which make up a multipart message.
+pub type FrameSet = Vec<Frame>;
+
 #[derive(Error, Debug)]
 pub enum MessageError {
     /// Multipart message is malformed
     MalformedMultipart,
+    /// Failed to serialize message
+    SerializeFailed,
     /// Failed to deserialize message
     DeserializeFailed,
-    // An error occurred serialising an object into binary
+    /// An error occurred serialising an object into binary
     BinarySerializeError,
-    // An error occurred deserialising binary data into an object
+    /// An error occurred deserialising binary data into an object
     BinaryDeserializeError,
 }
 
 bitflags! {
     #[derive(Deserialize, Serialize)]
-    struct IdentityFlags: u8 {
+    pub struct IdentityFlags: u8 {
         const ENCRYPTED = 0b00000001;
     }
 }
 
-#[derive(Debug, Deserialize, Serialize, PartialEq)]
-pub enum NodeDestination<PubKey>
-where PubKey: PublicKey
-{
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+pub enum NodeDestination<PubKey> {
     Unknown,
     PublicKey(PubKey),
     NodeId(NodeId),
 }
 
-#[derive(Debug, Deserialize, Serialize, PartialEq)]
-pub struct MessageEnvelopeHeader<PubKey>
-where PubKey: PublicKey
-{
-    version: u8,
-    source: PubKey,
-    dest: NodeDestination<PubKey>,
-    signature: Vec<u8>,
-    flags: IdentityFlags,
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+pub struct MessageEnvelopeHeader<PubKey> {
```

Comment:
It's actually a bit misleading to call your generic type `PubKey` here. There's no such constraint on the type. Maybe just calling it `P` is better since it gives a _hint_ that it should be more than just any `T`.

Reply:
Valid point, but on the flip side leaving that in sync with the impl makes the code easier to follow

Reply:
I agree it doesn't have to be a public key, but it does make the code more readable. This was already setup as PubKey from a previous PR, do you think I should change it?